### PR TITLE
Weld 3.1.0.Final

### PR DESCRIPTION
**Pull request by manovotn**

### CODE REVIEW

File: news/2019-02-07-weld-310Final.asciidoc [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+NOTE: Integrators (WildFly, Liberty, GlassFish, ...) will have to adjust to the changes made in SPI. Those are, most notably, the removal of long deprecated methods (link:https://issues.jboss.org/browse/WELD-2558[WELD-2558, window="_blank"]) and reworked `ProxyServices` which now delegate class defining responsibility to the integrator (link:https://issues.jboss.org/browse/WELD-2556[WELD-2556, window="_blank"]). This is crucial for Weld to operate on newer JDKs (12) and while Weld will work with the old approach for now, we will eventually fully swap to this new SPI. 
+
+Fixes and improvements:
+
+* Weld API/SPI
+** Added API to allow for CDI context propagation between threads (link:https://issues.jboss.org/browse/WELD-2497[WELD-2497, window="_blank"])
+*** New API class was introduced (`WeldAlterableContext`), this class offers methods to manipulate context state
+*** You can now propagate `Request`, `Session` and `Conversation` contexts between threads as they all implement `WeldAlterableContext`
+*** This comes with certain limitations is mostly designed for frameworks which would do this for you (such as MicroProfile Concurrency) but anyone can use it
```
Comment (nziakova) [inline news/2019-02-07-weld-310Final.asciidoc]: This comes with certain limitations is mostly designed for frameworks... - the wording does not look correct to me, or I just don't understand

File: news/2019-02-07-weld-310Final.asciidoc [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+** `WeldManager`, providing extra methods over what `BeanManager` offers, is now an injectable bean (link:https://issues.jboss.org/browse/WELD-2538[WELD-2538, window="_blank"])
+** `WeldManager` now offers new util methods allowing you to easily grab active contexts
+*** This is mainly for the purpose of context propagation as it also allows you to grab all active contexts that support it
+** `WeldManager` can now be used to check if any given context is active without having to care about exceptions (link:https://issues.jboss.org/browse/WELD-2537[WELD-2537, window="_blank"])
+** Added new SPI for class defining, deprecated old approach in `ProxyServices` (link:https://issues.jboss.org/browse/WELD-2556[WELD-2556, window="_blank"])
+*** Integrators will have to implement this new API to be able to operate on JDK 12+
+*** Weld now delegates class defining to integrators in order to avoid having to crack open `ClassLoader` methods
+*** Previously, integrators provided `ClassLoader` instance which Weld then used to invoke `defineClass()` methods
+*** We not ask integrators to invoke those methods themselves while providing them with all the necessary bits for doing so
```
Comment (nziakova) [inline news/2019-02-07-weld-310Final.asciidoc]: not -> now

File: news/2019-02-07-weld-310Final.asciidoc [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+** You can now use `InterceptionFactory` with an interface as a parameter (link:https://issues.jboss.org/browse/WELD-2550[WELD-2550, window="_blank"], link:https://issues.jboss.org/browse/WELD-2533[WELD-2533, window="_blank"])
+*** This means the proxy class will be based off an interface which is by definition always proxyable
+*** You can therefore even supply an unproxyable implementation and it will still work
+*** Note that this is experimental feature with some limitations to it, see link:http://docs.jboss.org/weld/reference/latest/en-US/html_single/#_loosening_the_limitations_of_literal_interceptionfactory_literal[Weld docs] for more details
+** Correct atomic behaviour in `RequestContextController` (link:https://issues.jboss.org/browse/WELD-2536[WELD-2536, window="_blank"])
+** Fix rare race condition in `ConcurrentValidator` (link:https://issues.jboss.org/browse/WELD-2545[WELD-2545, window="_blank"])
+** Small correction to interceptor resulution when they have no bindings (link:https://issues.jboss.org/browse/WELD-2521[WELD-2521, window="_blank"])
```
Comment (nziakova) [inline news/2019-02-07-weld-310Final.asciidoc]: resulution -> resolution

File: news/2019-02-07-weld-310Final.asciidoc [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+** Enforce consistent behaviour between `AnnotatedType` and `WithAnnotations` in regards to how default methods (link:https://issues.jboss.org/browse/WELD-2551[WELD-2551, window="_blank"])
```
Comment (nziakova) [inline news/2019-02-07-weld-310Final.asciidoc]: ... in regards to how default methods - the sentence looks unfinished

File: news/2019-02-07-weld-310Final.asciidoc [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+** All proxy-specific methods added by Weld now have more complex names to avoid (very rare) method clashes (link:https://issues.jboss.org/browse/WELD-2508[WELD-2508, window="_blank"])
+** Lower logging level of `InterceptorLogger.unableToDetermineInterceptedBean()` from `WARN` to `INFO` (link:https://issues.jboss.org/browse/WELD-2546[WELD-2546, window="_blank"])
+
+* Weld SE
+** You can now extend the set of bean defining annotations in Weld SE (link:https://issues.jboss.org/browse/WELD-2523[WELD-2523, window="_blank"])
+*** This is handy for when you have discovery on and want custom beans automatically picked up while still avoiding the `all` discovery mode
+** Fix bug in handling JAR dependencies added onto classpath where you could accidentally add more packages then desired (link:https://issues.jboss.org/browse/WELD-2535[WELD-2535, window="_blank"])
```
Comment (nziakova) [inline news/2019-02-07-weld-310Final.asciidoc]: then -> than

File: news/2019-02-07-weld-310Final.asciidoc [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+** Fix rare race condition in `ConcurrentValidator` (link:https://issues.jboss.org/browse/WELD-2545[WELD-2545, window="_blank"])
+** Small correction to interceptor resulution when they have no bindings (link:https://issues.jboss.org/browse/WELD-2521[WELD-2521, window="_blank"])
+** Enforce consistent behaviour between `AnnotatedType` and `WithAnnotations` in regards to how default methods (link:https://issues.jboss.org/browse/WELD-2551[WELD-2551, window="_blank"])
+** All proxy-specific methods added by Weld now have more complex names to avoid (very rare) method clashes (link:https://issues.jboss.org/browse/WELD-2508[WELD-2508, window="_blank"])
+** Lower logging level of `InterceptorLogger.unableToDetermineInterceptedBean()` from `WARN` to `INFO` (link:https://issues.jboss.org/browse/WELD-2546[WELD-2546, window="_blank"])
+
+* Weld SE
+** You can now extend the set of bean defining annotations in Weld SE (link:https://issues.jboss.org/browse/WELD-2523[WELD-2523, window="_blank"])
+*** This is handy for when you have discovery on and want custom beans automatically picked up while still avoiding the `all` discovery mode
```
Comment (nziakova) [inline news/2019-02-07-weld-310Final.asciidoc]: This is handy for when you have.. - is the "for" needed here?

File: news/2019-02-07-weld-310Final.asciidoc (hunk 4 above)
Comment (manovotn) [reply news/2019-02-07-weld-310Final.asciidoc]: just the `how` is redundant

File: news/2019-02-07-weld-310Final.asciidoc (hunk 2 above)
Comment (manovotn) [reply news/2019-02-07-weld-310Final.asciidoc]: thanks

File: news/2019-02-07-weld-310Final.asciidoc (hunk 6 above)
Comment (manovotn) [reply news/2019-02-07-weld-310Final.asciidoc]: Optional, should work both ways.

**Comment by manovotn** [comment]
Updated per comments, thanks for review!

**Pull request merged**